packages/ai/src/agent.ts

```diff
@@ -256,16 +264,26 @@ export class AIAgent extends BaseActor<AgentMetadata> implements IIAgent {
   }
 
   protected async _getOrCreateCallableToolSet(
-    mcpId: string,
+    connection: string | MCPConnection,
     signal?: AbortSignal,
   ): Promise<ToolsInput | null> {
-    if (this.callableToolSet[mcpId]) {
-      return this.callableToolSet[mcpId];
+    const isConnectionMcpId = typeof connection === "string";
+
+    if (isConnectionMcpId && this.callableToolSet[connection]) {
+      return this.callableToolSet[connection];
     }
 
-    const integration = await this.metadata?.mcpClient?.INTEGRATIONS_GET({
-      id: mcpId,
-    });
+    const mcpId = normalizeMCPId(connection);
+
+    const integration = isConnectionMcpId
+      ? await this.metadata?.mcpClient?.INTEGRATIONS_GET({
+        id: connection,
+      })
+      : {
+        connection,
+        name: mcpId,
+        id: mcpId,
+      };
```

Comment:
quick improvement:

const [mcpId, connection] = typeof connection === "string" ? [connection, await await this.metadata?.mcpClient?.INTEGRATIONS_GET({
        id: connection,
      })]: [normalizeMCPId(connection), connection]: 
      
      the rest of the code remains the same, you can cache callableToolSets too